File: assets/mods/std/section/movement.mjs
import { parse } from "package:smol-toml/index.js";
import { Synchronize } from "std:synchronize.mjs";

export const MovementType = {
  LAND: "Land",
};

export class Movement extends Synchronize {
  constructor(
    movementType,
    maxMoveSpeed,
    moveAcceleration,
    moveDeceleration,
    reversePercentage,
    maxTurnSpeed,
    turnAcceleration,
    turnDeceleration,
    angularDamping,
    linearDamping
  ) {
    super();
    this.movementType = movementType;
    this.maxMoveSpeed = maxMoveSpeed;
    this.moveAcceleration = moveAcceleration;
    this.moveDeceleration = moveDeceleration;
    this.reversePercentage = reversePercentage;
    this.maxTurnSpeed = maxTurnSpeed;
    this.turnAcceleration = turnAcceleration;
    this.turnDeceleration = turnDeceleration;
    this.angularDamping = angularDamping;
    this.linearDamping = linearDamping;
  }
}

export function fromValues(
  movementType,
  maxMoveSpeed,
  moveAcceleration,
  moveDeceleration,
  reversePercentage,
  maxTurnSpeed,
  turnAcceleration,
  turnDeceleration,
  angularDamping,
  linearDamping
) {
  movementType =
    typeof movementType !== "undefined" ? movementType : MovementType.LAND;
  maxMoveSpeed = typeof maxMoveSpeed !== "undefined" ? maxMoveSpeed : 0;
  moveAcceleration =
    typeof moveAcceleration !== "undefined" ? moveAcceleration : 0;
  moveDeceleration =
    typeof moveDeceleration !== "undefined" ? moveDeceleration : 0;
  reversePercentage =
    typeof reversePercentage !== "undefined" ? reversePercentage : 0;
  maxTurnSpeed = typeof maxTurnSpeed !== "undefined" ? maxTurnSpeed : 0;
  turnAcceleration =
    typeof turnAcceleration !== "undefined" ? turnAcceleration : 0;
  turnDeceleration =
    typeof turnDeceleration !== "undefined" ? turnDeceleration : 0;
  return new Movement(
    movementType,
    maxMoveSpeed,
    moveAcceleration,
    moveDeceleration,
    reversePercentage,
    maxTurnSpeed,
    turnAcceleration,
    turnDeceleration
  );
}

export function fromToml(toml) {
  let movement = parse(toml);
  movement.movementType =
    typeof movement.movementType !== "undefined"
      ? movement.movementType
      : MovementType.LAND;
  movement.maxMoveSpeed =
    typeof movement.maxMoveSpeed !== "undefined" ? movement.maxMoveSpeed : 0;
  movement.moveAcceleration =
    typeof movement.moveAcceleration !== "undefined"
      ? movement.moveAcceleration
      : 0;
  movement.moveDeceleration =
    typeof movement.moveDeceleration !== "undefined"
      ? movement.moveDeceleration
      : 0;
  movement.reversePercentage =
    typeof movement.reversePercentage !== "undefined"
      ? movement.reversePercentage
      : 0;
  movement.maxTurnSpeed =
    typeof movement.maxTurnSpeed !== "undefined" ? movement.maxTurnSpeed : 0;
  movement.turnAcceleration =
    typeof movement.turnAcceleration !== "undefined"
      ? movement.turnAcceleration
      : 0;
  movement.turnDeceleration =
    typeof movement.turnDeceleration !== "undefined"
      ? movement.turnDeceleration
      : 0;
  return new Movement(
    movement.movementType,
    movement.maxMoveSpeed,
    movement.moveAcceleration,
    movement.moveDeceleration,
    movement.reversePercentage,
    movement.maxTurnSpeed,
    movement.turnAcceleration,
    movement.turnDeceleration
  );
}

File: src/custom_unit/section/movement.rs
use crate::bevy_ext::try_from_js::*;
use avian2d::math::Scalar;
use bevy::prelude::*;
use boa_engine::{JsResult, prelude::*, value::TryFromJs};
#[derive(Debug, Default, Clone, Copy, Component, Reflect, TryFromJs)]
pub struct Movement {
    pub movement_type: MovementType,
    #[boa(from_js_with = "f32_try_from_js")]
    #[boa(rename = "maxMoveSpeed")]
    pub max_move_speed: Scalar,
    #[boa(from_js_with = "f32_try_from_js")]
    #[boa(rename = "moveAcceleration")]
    pub move_acceleration: Scalar,
    #[boa(from_js_with = "f32_try_from_js")]
    #[boa(rename = "moveDeceleration")]
    pub move_deceleration: Scalar,
    #[boa(from_js_with = "f32_try_from_js")]
    #[boa(rename = "reversePercentage")]
    pub reverse_percentage: Scalar,
    #[boa(from_js_with = "f32_try_from_js")]
    #[boa(rename = "maxTurnSpeed")]
    pub max_turn_speed: Scalar,
    #[boa(from_js_with = "f32_try_from_js")]
    #[boa(rename = "turnAcceleration")]
    pub turn_acceleration: Scalar,
    #[boa(from_js_with = "f32_try_from_js")]
    #[boa(rename = "turnDeceleration")]
    pub turn_deceleration: Scalar,
}

impl Movement {
    pub fn new(
        movement_type: MovementType,
        max_move_speed: Scalar,
        move_acceleration: Scalar,
        move_deceleration: Scalar,
        reverse_percentage: Scalar,
        max_turn_speed: Scalar,
        turn_acceleration: Scalar,
        turn_deceleration: Scalar,
    ) -> Self {
        Self {
            movement_type,
            max_move_speed,
            move_acceleration,
            move_deceleration,
            reverse_percentage,
            max_turn_speed,
            turn_acceleration,
            turn_deceleration,
        }
    }
}

#[derive(Debug, Default, Clone, Copy, Component, Reflect)]
pub enum MovementType {
    #[default]
    Land,
}

impl TryFromJs for MovementType {
    fn try_from_js(value: &JsValue, _context: &mut Context) -> JsResult<Self> {
        match value {
            JsValue::String(movement_type) => match movement_type.to_std_string_lossy().as_str() {
                "Land" => Ok(MovementType::Land),
                _ => Err(JsNativeError::typ()
                    .with_message("cannot convert value to a movement_type")
                    .into()),
            },
            _ => Err(JsNativeError::typ()
                .with_message("cannot convert value to a movement_type")
                .into()),
        }
    }
}
